pytensor/tensor/utils.py

```
@@ -202,6 +202,39 @@ def _parse_gufunc_signature(
     )
 
 
+def _gufunc_to_out_shape(
```

This should try to prioritize input dimensions that are constant when there are multiple ones with the same letter, as it will generate a better shape graph.

Also fail explicitly if the shape can't be inferred from the signature alone

---

Should I raise a `ValueError` when a single dimension in the `gufunc_signature`  (say `m`) is assigned two values in the input `shapes`? I am raising a `ValueError` when the shape cannot be inferred from the signature. Here is my new function:
```python
def _gufunc_to_out_shape(
    signature: str, shapes: list[tuple[int, ...]]
) -> list[tuple[int, ...]]:
    """
    Compute the shape of the output of an Op given its gufunc signature and the
    shapes of its inputs.

    Parameters
    ----------
    signature : str
        The gufunc signature of the Op.
        eg: "(m,n),(n,p)->(m,p)".

    shapes : list of tuple of int
        The list of shapes of the inputs.

    Returns
    -------
    out_shape : list of tuple of int
        The list of shapes of the outputs.

    Raises
    ------
    ValueError
        If the signature is invalid for the shapes of the inputs.
    """
    parsed = _parse_gufunc_signature(signature)
    out_shape = []
    dim_to_size = dict()
    for i, input_shape in enumerate(parsed[0]):
        for j, dim in enumerate(input_shape):
            if dim not in dim_to_size:
                dim_to_size[dim] = shapes[i][j]
            elif dim_to_size[dim] != shapes[i][j]:
                raise ValueError(
                    f"Invalid signature {signature} for shapes {shapes}."
                    f"Dimension {dim} at is not consistent across inputs."
                )

    for i, output_shape in enumerate(parsed[1]):
        temp_list = []
        for j, dim in enumerate(output_shape):
            if dim not in dim_to_size:
                raise ValueError(
                    f"Invalid signature {signature} for shapes {shapes}. "
                    f"Dimension {dim} not in input dimensions."
                )
            else:
                temp_list.append(dim_to_size[dim])
        out_shape.append(tuple(temp_list))
    return out_shape
```

---

Okay, so if `shapes` list can contain symbolic values, can we first convert the symbolic variable to a `tensor_variable` using `as_tensor_variables` and then use `.equals` method to compare them? Specifically, in the function above, we make this change:
```python
for i, input_shape in enumerate(parsed[0]):
    for j, dim in enumerate(input_shape):
        current_dim = as_tensor_variable(shapes[i][j])
        if dim not in dim_to_size:
            dim_to_size[dim] = current_dim
        elif not dim_to_size[dim].equals(current_dim):
            raise ValueError(
                f"Invalid signature {signature} for shapes {shapes}. "
                f"Dimension {dim} is not consistent across inputs."
            )
```
 This is how the output looks like then: (The shapes input contains integers here)
```python
print(_gufunc_to_out_shape("(m,m)->(m),(m,m)", [(2, 2)]))
[(TensorConstant(TensorType(int8, shape=()), data=array(2, dtype=int8)),), (TensorConstant(TensorType(int8, shape=()), data=array(2, dtype=int8)), TensorConstant(TensorType(int8, shape=()), data=array(2, dtype=int8)))]
```
Can there be a better way to do this?

---

No, if they are symbolic, unless they are constants, you can't know if they are equivalent, so raising an error is incorrect. User may have defined `x = pt.vector("x"); y =pt.vector("y")`; out = x + y`. You will not know if x and y are identical until the user compiles a function and provides values for them.

The logic for this is already in the Blockwise infer_shape, you should be able to just grab it and refactor it.

---

I took the code from the `infer_shape` of `Blockwise` class, and used it to write the function. This gives priority to `Constants`. Is the output type `list[tuple[Any, ...]]` fine?
```python
def _gufunc_to_out_shape(
    signature: str, shapes
) -> list[tuple[Any, ...]]:
    """
    Compute the shape of the output of an Op given its gufunc signature and the
    shapes of its inputs.

    Parameters
    ----------
    signature : str
        The gufunc signature of the Op.
        eg: "(m,n),(n,p)->(m,p)".

    shapes : list of tuple of Any
        The list of shapes of the inputs.

    Returns
    -------
    out_shape : list of tuple of Any
        The list of shapes of the outputs.

    Raises
    ------
    ValueError
        If the signature is invalid for the shapes of the inputs.
    """
    input_sig, output_sig = _parse_gufunc_signature(signature)
    dim_to_size : dict[str, Any] = {}
    for input_shape, sig in zip(shapes, input_sig, strict = True):
        for size, dim_name in zip(input_shape, sig, strict=True):
            prev_size = dim_to_size.get(dim_name)
            if prev_size is None:
                dim_to_size[dim_name] = size
            # Prefer constants
            elif not isinstance(prev_size, Constant):
                dim_to_size[dim_name] = size
            elif prev_size.data != size:
                raise ValueError(
                    f"Invalid signature {signature} for shapes {shapes}. "
                    f"Dimension {dim_name} is not consistent across inputs."
                )
    out_shapes = []
    for output_shape in output_sig:
        temp_list = []
        for dim in output_shape:
            if dim not in dim_to_size:
                raise ValueError(
                    f"Invalid signature {signature} for shapes {shapes}. "
                    f"Dimension {dim} not in input dimensions."
                )
            else:
                temp_list.append(dim_to_size[dim])
        out_shapes.append((*temp_list,))       
    return out_shapes
    
 ```